zproject/backends.py

```diff
@@ -902,9 +930,11 @@ def get_or_build_user(
                 )
                 do_reactivate_user(user)
 
+        self.sync_mail_from_ldap(username, ldap_user, user.realm)
```

Comment:
I think this `sync_ldap_user` codepath looks wrong, have you tested it? If the user email has changed, then the `get_user_by_delivery_email` above will have thrown an exception - so we're not handling things correctly with this approach :thinking: The syncing codepath in general seems tricky to figure out with this feature. Perhaps the first right thing to do would be to change the lookup in sync_user_from_ldap:
```
ldap_username = backend.django_to_ldap_username(user_profile.delivery_email)
```
to lookup by DN if the `user_profile` has a dn saved on it? Not sure

But in the current shape of that code,
```
def sync_user_from_ldap(user_profile: UserProfile, logger: logging.Logger) -> bool:
    backend = ZulipLDAPUserPopulator()
    try:
        ldap_username = backend.django_to_ldap_username(user_profile.delivery_email)
    except ZulipLDAPExceptionNoMatchingLDAPUser:
        if (
            settings.ONLY_LDAP
            if settings.LDAP_DEACTIVATE_NON_MATCHING_USERS is None
            else settings.LDAP_DEACTIVATE_NON_MATCHING_USERS
        ):
            do_deactivate_user(user_profile)
            logger.info("Deactivated non-matching user: %s", user_profile.delivery_email)
            return True
        elif user_profile.is_active:
            logger.warning("Did not find %s in LDAP.", user_profile.delivery_email)
        return False
```

I think the email lookup will fail and the user will get deactivated?